community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/index/DocValuesCollector.java

```diff
@@ -0,0 +1,852 @@
+ * segment that contains a hit. Those items can be later used to read {@code DocValues} fields
+ * and iterate over the matched {@code DocIdSet}s. This collector is different from
+ * {@code org.apache.lucene.search.CachingCollector} in that the later focuses on predictable RAM usage
+ * and feeding other collectors while this collector focuses on exposing the required per-segment data structures
+ * to the user.
+ */
+public class DocValuesCollector extends SimpleCollector
+{
+
+    private LeafReaderContext context;
+    private int segmentHits;
+    private int totalHits;
+    private Scorer scorer;
+    private float[] scores;
+    private final boolean keepScores;
+    private final List<MatchingDocs> matchingDocs = new ArrayList<>();
+    private Docs docs;
+
+    /**
+     * Default Constructor, does not keep scores.
+     */
+    public DocValuesCollector()
+    {
+        this( false );
+    }
+
+    /**
+     * @param keepScores true if you want to trade correctness for speed
+     */
+    public DocValuesCollector( boolean keepScores )
+    {
+        this.keepScores = keepScores;
+    }
+
+    /**
+     * @param field the field that contains the values
+     * @return an iterator over all NumericDocValues from the given field
+     */
+    public LongValuesIterator getValuesIterator( String field )
+    {
+        return new LongValuesIterator( getMatchingDocs(), getTotalHits(), field );
+    }
+
+    /**
+     * @param field the field that contains the values
+     * @param sort how the results should be sorted
+     * @return an iterator over all NumericDocValues from the given field with respect to the given sort
+     * @throws IOException
+     */
+    public PrimitiveLongIterator getSortedValuesIterator( String field, Sort sort ) throws IOException
+    {
+        if ( sort == null || sort == Sort.INDEXORDER )
+        {
+            return getValuesIterator( field );
+        }
+        int size = getTotalHits();
+        if ( size == 0 )
+        {
+            return PrimitiveLongCollections.emptyIterator();
+        }
+        TopDocs topDocs = getTopDocs( sort, size );
+        LeafReaderContext[] contexts = getLeafReaderContexts( getMatchingDocs() );
+        return new TopDocsValuesIterator( topDocs, contexts, field );
+    }
+
+    /**
+     * Replay the search and collect every hit into TopDocs. One {@code ScoreDoc} is allocated
+     * for every hit and the {@code Document} instance is loaded lazily with on every iteration step.
+     *
+     * @param sort how to sort the iterator. If this is null, results will be in index-order.
+     * @return an indexhits iterator over all matches
+     * @throws IOException
+     */
+    public IndexHits<Document> getIndexHits( Sort sort ) throws IOException
+    {
+        List<MatchingDocs> matchingDocs = getMatchingDocs();
+        int size = getTotalHits();
+        if ( size == 0 )
+        {
+            return new AbstractIndexHits<Document>()
```

Comment:
This should be a static singleton instead of a new instance every time... there should already be something like this somewhere in the code base.